README.rst
============
 OpenSlides
============

I. What is OpenSlides?
======================

OpenSlides is a free web-based presentation and assembly system for
displaying and controlling of agenda, motions and elections of an assembly.
See http://openslides.org/ for more information.


II. Requirements
================

OpenSlides runs everywhere where Python is running (for example on
GNU/Linux, Mac or Windows (XP or newer)). On each client you need only an
actual webbrowser.


III. Installation
=================

Installation on GNU/Linux or Mac OS X
-------------------------------------

1. Check requirements

   Make sure that you have installed Python Programming Language 2 (>= 2.6)
   on your system. You will also need the Python development headers.

   For example for Ubuntu run::

       $ sudo apt-get install python-dev

2. Setup a virtual environment with Virtual Python Environment builder
   (optional)

   You can setup a virtual environment to install OpenSlides as non-root
   user. Make sure that you have installed Virtual Python Environment
   builder on your system.

   For example for Ubuntu run::

       $ sudo apt-get install python-virtualenv

   Create your OpenSlides directory, change to it, setup and activate the
   virtual environment::

       $ mkdir OpenSlides
       $ cd OpenSlides
       $ virtualenv .virtualenv
       $ source .virtualenv/bin/activate

3. Install OpenSlides

   To use the Python Package Index (PyPI) simply run::

       $ pip install openslides

   You can also use the package from the `OpenSlides Website
   <http://openslides.org/download/>`_. Download latest OpenSlides release
   as compressed tar archive and run::

       $ pip install openslides-x.x.tar.gz

   OpenSlides will install all required python packages (see
   requirements_production.txt).


Installation on Windows
-----------------------

*Note: There is a portable version of OpenSlides for Windows which does not
required any install steps. If there is a reason that you can not use the
portable version you should observe the following install steps.*

1. Check requirements

   Make sure that you have installed Python Programming Language 2 (>= 2.6)
   and Setuptools on your system.

   a. Download and run the `Python 32-bit MSI installer
      <http://www.python.org/ftp/python/2.7.5/python-2.7.5.msi>`_. Note
      that the 32-bit MSI installer is required even on a 64-bit Windows
      system. If you use the 64-bit MSI installer, step 3 of this
      instruction will fail unless you installed the packages reportlab and
      pillow manually.

   b. Add python directories to PATH (via Control Panel > System >
      Advanced): ``";C:\\Python27;C:\\Python27\\Scripts"``. Note that the path
      can differ if you customized the install of Python in the first step.

   c. Download and run (via double click) the last `install script
      ez_setup.py for Setuptools
      <https://pypi.python.org/pypi/setuptools/#windows>`_.

2. Setup a virtual environment with Virtual Python Environment builder
   (optional)

   You can setup a virtual environment to install OpenSlides as non-root
   user. Make sure that you have installed Virtual Python Environment
   builder on your system.

   To install Virtual Python Environment builder, open command line (cmd)
   and run::

       > easy_install https://pypi.python.org/packages/source/v/virtualenv/virtualenv-1.10.1.tar.gz

   Create your OpenSlides directory, change to it, setup and activate the
   virtual environment::

       > md OpenSlides
       > cd OpenSlides
       > virtualenv .virtualenv
       > .virtualenv\Scripts\activate

3. Install OpenSlides

   To use the Python Package Index (PyPI) simply run on command line (cmd)::

       > easy_install openslides

   You can also use the package from the `OpenSlides Website
   <http://openslides.org/download/>`_. Download latest OpenSlides release
   as compressed tar archive and run::

       > easy_install openslides-x.x.tar.gz

   OpenSlides will install all required python packages (see
   requirements_production.txt).


IV. Start
=========

To start OpenSlides simply run on command line::

    openslides

If you run this command the first time, a new database and the admin account
(Username: `admin`) will be created. Please change the password (Password:
`admin`) after first login!

OpenSlides will start using the integrated Tornado webserver. It will also
try to open the webinterface in your default webbrowser. The server will
try to listen on the local ip address on port 80 or port 8000 if you do not
have admin permissions. That means that the server will be available to
everyone on your local network (at least for commonly used network
configurations).

If you use a virtual environment (see install instructions, step 2), do not
forget to activate the environment before restart after you have closed the
terminal.

For Unix and Mac OS X run::

    $ source .virtualenv/bin/activate

For Windows run::

    > .virtualenv\Scripts\activate

To get help on the command line options run::

    openslides --help


V. Development
==============

If you want to join us developing OpenSlides, have a look at `GitHub
<https://github.com/OpenSlides/OpenSlides/>`_ or write an email to our
`mailing list <http://openslides.org/contact/>`_.


Installation and start of the development version
-------------------------------------------------

1. Check requirements

   Follow step 1 in the correspondent instruction in section III.

2. Get OpenSlides source code

   Clone current master version from `OpenSlides' GitHub repository
   <https://github.com/OpenSlides/OpenSlides>`_. This requires `Git
   <http://git-scm.com/>`_.

   For example for Ubuntu run::

       $ sudo apt-get install git
       $ git clone git://github.com/OpenSlides/OpenSlides.git
       $ cd OpenSlides

   For Windows you can use GitBash::

       > cd ...  # Go to a nice place in your filesystem.
       > git clone git://github.com/OpenSlides/OpenSlides.git
       > cd OpenSlides

3. Setup a virtual environment with Virtual Python Environment builder (optional)

   Follow step 2 in the correspondent instruction in section III.

4. Install all required python packages

   For Unix and Mac OS X run::

       $ pip install -r requirements.txt

   For Windows run::

       > easy_install  # Insert all packages from requirements.txt and requirements_production.txt here

5. Start OpenSlides

   To start OpenSlides use the command line script::

       python manage.py start

   To get help on the command line options run::

       python manage.py --help


Coding Style
------------

You can find some information on the coding style in the `OpenSlides wiki
<https://github.com/OpenSlides/OpenSlides/wiki/De%3ACode-Richtlinien-f%C3%BCr
-Openslides>`_.


VI. Used software
=================

OpenSlides uses the following projects or parts of them:

* `Beautiful Soup <http://www.crummy.com/software/BeautifulSoup/>`_,
  License: MIT

* `Bleach <https://github.com/jsocol/bleach/>`_, License: BSD

* `CKEditor <http://ckeditor.com>`_, License: GPL, LGPL, MPL

* `Django <https://www.djangoproject.com>`_, License: BSD

* `Django mptt <https://github.com/django-mptt/django-mptt/>`_, License: BSD

* `Django haystack <http://haystacksearch.org>`_, License: BSD

* `pdf.js <http://mozilla.github.io/pdf.js/>`_, License: Apache License v2.0

* `Pillow <https://github.com/python-imaging/Pillow/>`_, License: Standard
  PIL License

* `qrcode <https://github.com/lincolnloop/python-qrcode/>`_, License: BSD

* `ReportLab <http://www.reportlab.com/software/opensource/rl-toolkit/>`_,
  License: BSD

* `sockjs-client <https://github.com/sockjs/sockjs-client>`_,
  License: MIT

* `sockjs-tornado <https://github.com/mrjoes/sockjs-tornado>`_,
  License: MIT

* `Sphinx <http://sphinx-doc.org/>`_, License: BSD

* `Sphinx Bootstrap Theme
  <http://ryan-roemer.github.io/sphinx-bootstrap-theme/>`_, License: MIT

* `Tornado <http://www.tornadoweb.org/en/stable/>`_, License: Apache
  License v2.0

* `Twitter Bootstrap <http://getbootstrap.com/2.3.2/>`_, License: Apache
  License v2.0

* `Ubuntu TrueType Font <http://font.ubuntu.com>`_, License: Ubuntu Font
  Licence 1.0

* `Whoosh <https://bitbucket.org/mchaput/whoosh/wiki/Home/>`_, License: BSD

* `jQuery <http://www.jquery.com>`_, License: MIT

* jQuery Plugins:

  - `jQuery DataTables Plugin <http://www.datatables.net>`_, License:
    BSD/GPLv2

  - `jQuery Cookie Plugin <https://github.com/carhartl/jquery-cookie/>`_,
    License: MIT/GPL

  - `jQuery Form Plugin <http://malsup.com/jquery/form/>`_, License: MIT/GPLv2

  - `jQuery Once Plugin <http://plugins.jquery.com/once/>`_, License: MIT/GPL

  - `jQuery Templating Plugin
    <https://github.com/BorisMoore/jquery-tmpl/>`_, License: MIT/GPLv2

  - `jQuery bsmSelect <https://github.com/vicb/bsmSelect/>`_, License:
    MIT/GPLv2

* `jQuery UI <http://jqueryui.com>`_ with custom ui components: core,
  widget, mouse, sortable, datepicker, slider and css theme 'smoothness',
  License: MIT

* jQuery UI AddOns:

  - `jQuery UI Nested Sortable
    <https://github.com/mjsarfatti/nestedSortable/>`_, License: MIT

  - `jQuery UI Slider Access
    <http://trentrichardson.com/examples/jQuery-SliderAccess/>`_, License:
    MIT/GPLv2

  - `jQuery timepicker addon
    <http://trentrichardson.com/examples/timepicker/>`_, License: MIT/GPLv2


VII. License and authors
========================

OpenSlides is Free/Libre Open Source Software (FLOSS), and distributed
under the MIT License, see LICENSE file. The authors of OpenSlides are
mentioned in the AUTHORS file.


CHANGELOG
=========================
 CHANGELOG of OpenSlides
=========================

https://openslides.org/

Version 2.1 (unreleased)
========================
[https://github.com/OpenSlides/OpenSlides/milestones/2.1]

Agenda:
- Added button to remove all speakers from a list of speakers.
- Added option to create or edit agenda items as subitems of others.

Assignments:
- Remove unused assignment config to publish winner election results only.
- Added options to calculate percentages on different bases.

Core:
- Added support for big assemblies with lots of users.
- Added HTML support for messages on the projector.
- Moved custom slides to own app "topics". Renamed it to "Topic".

Motions:
- Added origin field.
- Added button to sort and number all motions in a category.
- Introduced pdfMake for clientside generation of PDFs.
- Added configurable fields for comments.
- Added recommendations for motions.
- Changed label of former state "commited a bill" to "refered to committee".
- Added options to calculate percentages on different bases.

Users:
- Added field is_committee and new default group Committees.
- Added field number.
- Added new matrix-interface for managing groups and their permissions.

Other:
- Django 1.10 is now supported. Dropped support for Django 1.8 and 1.9.
- Removed config cache to support multiple threads or processes.
- Fixed bug, that the last change of a config value was not send via autoupdate.
- Added template hooks for plugins (in item detail view and motion poll form).
- Used Django Channels instead of Tornado. Refactoring of the autoupdate process.
- Added new caching system with support for Redis.


Version 2.0 (2016-04-18)
========================
[https://github.com/OpenSlides/OpenSlides/milestones/2.0]

OpenSlides 2.0 is essentially not compatible to OpenSlides 1.7. E. g.
customized templates, databases and plugins can not be reused without
adaption.

Agenda:
- Updated the tests and changed internal parts of method of the agenda model.
- Changed API of related objects. All assignments, motions and custom slides
  are now agenda items and can be hidden.
- Removed django-mptt.
- Added attachments to custom sldies.
- Improved CSV import.
Assignments:
- Renamed app from assignment to assignments.
- Removed possibility to block candidates.
- Massive refactoring and cleanup of the app.
Motions:
- Renamed app from motion to motions.
- Massive refactoring and cleanup of the app.
Mediafiles:
- Renamed app from mediafile to mediafiles.
- Used improved pdf presentation with angular-pdf.
- Massive refactoring and cleanup of the app.
Users:
- Massive refactoring of the participant app. Now called 'users'.
- Used new anonymous user object instead of an authentification backend. Used
  special authentication class for REST requests.
- Used authentication frontend via AngularJS.
- Improved CSV import.
Other:
- New OpenSlides logo.
- New design for web interface.
- Added multiple countdown support.
- Added colored countdown for the last n seconds (configurable).
- Switched editor from CKEditor to TinyMCE.
- Changed supported Python version to >= 3.4.
- Used Django 1.8 as lowest requirement.
- Django 1.9 is supported
- Added Django's application configuration. Refactored loading of signals
  and projector elements/slides.
- Setup migrations.
- Added API using Django REST Framework 3.x. Added several views and mixins
  for generic Django REST Framework views in OpenSlides apps.
- Removed most of the Django views and templates.
- Removed Django error pages.
- Added page for legal notice.
- Refactored projector API using metaclasses now.
- Renamed SignalConnectMetaClass classmethod get_all_objects to get_all
  (private API).
- Refactored config API and moved it into the core app.
- Removed old style personal info page, main menu entries and widget API.
- Used AngularJS with additional libraries for single page frontend.
- Removed use of 'django.views.i18n.javascript_catalog'. Used angular-gettext
  now.
- Updated to Bootstrap 3.
- Used SockJS for automatic update of AngularJS driven single page frontend.
- Refactored plugin API.
- Refactored start script and management commands. Changed command line
  option and path for local installation.
- Refactored tests.
- Used Bower and gulp to manage third party JavaScript and Cascading Style
  Sheets libraries.
- Used setup.cfg for development tools.
- Removed code for documentation and for Windows portable version with GUI.
  Used new repositories for this. Cleaned up main repository.
- Updated all dependencies.
Translations:
- Updated DE, FR, CS and PT translations.
- Added ES translations.


Version 1.7 (2015-02-16)
========================
[https://github.com/OpenSlides/OpenSlides/milestones/1.7]

Core:
- New feature to tag motions, agenda and assignments.
- Fixed search index problem to index contents of many-to-many tables
  (e. g. tags of a motion).
- Fixed AttributeError in chatbox on_open method.
Motions:
- New Feature to create amendments, which are related to a parent motion.
- Added possibility to hide motions from non staff users in some states.
Assignments:
- Fixed permissions to alter assignment polls.
Other:
- Cleaned up utils.views to increase performance when fetching single objects
  from the database for a view (#1378).
- Fixed bug on projector which was not updated when an object was deleted.
- Fixed bug and show special characters in PDF like ampersand (#1415).
- Updated pdf.js to 1.0.907.
- Improve the usage of bsmselect jquery plugin.
- Updated translations.


Version 1.6.1 (2014-12-08)
==========================
[https://github.com/OpenSlides/OpenSlides/issues?milestone=16]

Agenda:
- Fixed error in item numbers.
Motions:
- Show supporters on motion slide if available.
- Fixed motion detail view template. Added block to enable extra content via
  plugins.
Assignments:
- Fixed PDF build error when an election has more than 20 posts or candidates.
Participants:
- Fixed participant csv import with group ids:

openslides/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    openslides.main
    ~~~~~~~~~~~~~~~

    Main script to start and set up OpenSlides.

    :copyright: 2011–2013 by OpenSlides team, see AUTHORS.
    :license: GNU GPL, see LICENSE for more details.
"""

import base64
import ctypes
import optparse
import os
import socket
import sys
import tempfile
import threading
import time
import webbrowser

from django.conf import ENVIRONMENT_VARIABLE
from django.core.management import execute_from_command_line

from openslides import get_version
from openslides.utils.tornado_webserver import run_tornado


CONFIG_TEMPLATE = """#!/usr/bin/env python
# -*- coding: utf-8 -*-

import openslides.main
from openslides.global_settings import *

# Use 'DEBUG = True' to get more details for server errors
# (Default for releases: 'False')
DEBUG = False
TEMPLATE_DEBUG = DEBUG

DBPATH = %(dbpath)s

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': DBPATH,
        'USER': '',
        'PASSWORD': '',
        'HOST': '',
        'PORT': '',
    }
}

# Set timezone
TIME_ZONE = 'Europe/Berlin'

# Make this unique, and don't share it with anybody.
SECRET_KEY = %(default_key)r

# Add OpenSlides plugins to this list (see example entry in comment)
INSTALLED_PLUGINS = (
#    'pluginname',
)

INSTALLED_APPS += INSTALLED_PLUGINS

# Absolute path to the directory that holds media.
# Example: "/home/media/media.lawrence.com/"
MEDIA_ROOT = %(media_root_path)s

"""

KEY_LENGTH = 30

# sentinel used to signal that the database ought to be stored
# relative to the portable's directory
_portable_db_path = object()


def process_options(argv=None, manage_runserver=False):
    if argv is None:
        argv = sys.argv[1:]

    parser = optparse.OptionParser(
        description="Run openslides using the tornado webserver")
    parser.add_option("-a", "--address", help="IP Address to listen on.")
    parser.add_option("-p", "--port", type="int", help="Port to listen on.")
    parser.add_option(
        "--syncdb", action="store_true",
        help="Update/create database before starting the server.")
    parser.add_option(
        "--backupdb", action="store", metavar="BACKUP_PATH",
        help="Make a backup copy of the database to BACKUP_PATH")
    parser.add_option(
        "--reset-admin", action="store_true",
        help="Make sure the user 'admin' exists and uses 'admin' as password.")
    parser.add_option(
        "-s", "--settings", help="Path to the openslides configuration.")
    parser.add_option(
        "--no-reload", action="store_true",
        help="Do not reload the development server.")
    parser.add_option(
        "--no-browser",
        action="store_false", dest="start_browser", default=True,
        help="Do not automatically start web browser.")
    parser.add_option(
        "--no-run", action="store_true",
        help="Do not start the development server.")
    parser.add_option(
        "--version", action="store_true",
        help="Show version and exit.")

    opts, args = parser.parse_args(argv)

    # Do not parse any argv if the script is started via manage.py runserver.
    # This simulates the django runserver command
    if manage_runserver:
        opts.start_browser = False
        opts.no_reload = False
        return opts

    if opts.version:
        print get_version()
        exit(0)

    if args:
        sys.stderr.write("This command does not take arguments!\n\n")
        parser.print_help()
        sys.exit(1)

    return opts


def main(argv=None, manage_runserver=False):
    opts = process_options(argv, manage_runserver)
    _main(opts)


def win32_portable_main(argv=None):
    """special entry point for the win32 portable version"""

    opts = process_options(argv)

    database_path = None

    if opts.settings is None:
        portable_dir = get_portable_path()
        try:
            fd, test_file = tempfile.mkstemp(dir=portable_dir)
        except OSError:
            portable_dir_writeable = False
        else:
            portable_dir_writeable = True
            os.close(fd)
            os.unlink(test_file)

        if portable_dir_writeable:
            opts.settings = os.path.join(
                portable_dir, "openslides", "settings.py")
            database_path = _portable_db_path

    _main(opts, database_path=database_path)


def _main(opts, database_path=None):
    # Find the path to the settings
    settings_path = opts.settings
    if settings_path is None:
        settings_path = get_user_config_path('openslides', 'settings.py')

    # Create settings if necessary
    if not os.path.exists(settings_path):
        create_settings(settings_path, database_path)

    # Set the django environment to the settings
    setup_django_environment(settings_path)

    # Find url to openslides
    addr, port = detect_listen_opts(opts.address, opts.port)


    # Create Database if necessary
    if not database_exists() or opts.syncdb:
        run_syncdb()
        ## set_system_url(url)
        create_or_reset_admin_user()

    # Reset Admin
    elif opts.reset_admin:
        create_or_reset_admin_user()

    if opts.backupdb:
        backup_database(opts.backupdb)

    if opts.no_run:
        return

    # Start OpenSlides
    reload = True
    if opts.no_reload:
        reload = False

    if opts.start_browser:
        if port == 80:
            suffix = ""
        else:
            suffix = ":%d" % port
        start_browser("http://localhost%s" % suffix)

    # Start the server
    run_tornado(addr, port, reload)


def create_settings(settings_path, database_path=None):
    settings_module = os.path.dirname(settings_path)

    if database_path is _portable_db_path:
        database_path = get_portable_db_path()
        dbpath_value = 'openslides.main.get_portable_db_path()'
        media_root_path_value = 'openslides.main.get_portable_media_root_path()'
    else:
        if database_path is None:
            database_path = get_user_data_path('openslides', 'database.sqlite')
        dbpath_value = repr(fs2unicode(database_path))
        media_root_path_value = repr(fs2unicode(get_user_data_path('openslides', 'media', '')))

    settings_content = CONFIG_TEMPLATE % dict(
        default_key=base64.b64encode(os.urandom(KEY_LENGTH)),
        dbpath=dbpath_value,
        media_root_path=media_root_path_value)

    if not os.path.exists(settings_module):
        os.makedirs(settings_module)

    if not os.path.exists(os.path.dirname(database_path)):
        os.makedirs(os.path.dirname(database_path))

    with open(settings_path, 'w') as file:
        file.write(settings_content)


def setup_django_environment(settings_path):
    settings_file = os.path.basename(settings_path)
    settings_module_name = "".join(settings_file.split('.')[:-1])
    if '.' in settings_module_name:
        print "'.' is not an allowed character in the settings-file"
        sys.exit(1)
    settings_module_dir = os.path.dirname(settings_path)
    sys.path.append(settings_module_dir)
    os.environ[ENVIRONMENT_VARIABLE] = '%s' % settings_module_name


def detect_listen_opts(address=None, port=None):
    if address is None:
        address = "0.0.0.0"

    if port is None:
        # test if we can use port 80
        s = socket.socket()
        port = 80
        try:
            s.bind((address, port))
            s.listen(-1)
        except socket.error:
            port = 8000
        finally:
            s.close()

    return address, port


def database_exists():
    """Detect if database exists"""
    # can't be imported in global scope as they already require
    # the settings module during import
    from django.db import DatabaseError
    from django.core.exceptions import ImproperlyConfigured
    from openslides.participant.models import User

    try:
        # TODO: Use another model, the User could be deactivated
        User.objects.count()
    except DatabaseError:
        return False
    except ImproperlyConfigured:
        print "Your settings file seems broken"
        sys.exit(0)
    else:
        return True


def run_syncdb():
    # now initialize the database
    argv = ["", "syncdb", "--noinput"]
    execute_from_command_line(argv)


def set_system_url(url):
    # can't be imported in global scope as it already requires
    # the settings module during import
    from openslides.config.api import config

    if config['participant_pdf_system_url'] == 'http://example.com:8000':
        config['participant_pdf_system_url'] = url


def create_or_reset_admin_user():
    # can't be imported in global scope as it already requires
    # the settings module during import
    from openslides.participant.models import User
    try:
        admin = User.objects.get(username="admin")
        print("Password for user admin was reset to 'admin'")
    except User.DoesNotExist:
        admin = User()
        admin.username = 'admin'
        admin.last_name = 'Administrator'
        print("Created default admin user")

    admin.is_superuser = True
    admin.default_password = 'admin'
    admin.set_password(admin.default_password)
    admin.save()


def backup_database(dest_path):
    argv = ["", "backupdb", "--destination={0}".format(dest_path)]
    execute_from_command_line(argv)


def start_browser(url):
    browser = webbrowser.get()

    def f():
        time.sleep(1)
        browser.open(url)

    t = threading.Thread(target=f)
    t.start()


def fs2unicode(s):
    if isinstance(s, unicode):
        return s
    fs_encoding = sys.getfilesystemencoding() or sys.getdefaultencoding()
    return s.decode(fs_encoding)


def get_user_config_path(*args):
    if sys.platform == "win32":
        return win32_get_app_data_path(*args)

    config_home = os.environ.get(
        'XDG_CONFIG_HOME', os.path.join(os.path.expanduser('~'), '.config'))

    return os.path.join(fs2unicode(config_home), *args)


def get_user_data_path(*args):
    if sys.platform == "win32":
        return win32_get_app_data_path(*args)

    data_home = os.environ.get(
        'XDG_DATA_HOME', os.path.join(
            os.path.expanduser('~'), '.local', 'share'))

    return os.path.join(fs2unicode(data_home), *args)


def is_portable():
    exename = os.path.basename(sys.executable).lower()
    return exename == "openslides.exe"


def get_portable_path(*args):
    # NOTE: sys.executable will be the path to openslides.exe
    #       since it is essentially a small wrapper that embeds the
    #       python interpreter

    if not is_portable():
        raise Exception(
            "Cannot determine portable path when "
            "not running as portable")

    portable_dir = fs2unicode(os.path.dirname(os.path.abspath(sys.executable)))
    return os.path.join(portable_dir, *args)


def get_portable_db_path():
    return get_portable_path('openslides', 'database.sqlite')


def get_portable_media_root_path():
    return get_portable_path('openslides', 'media', '')


def win32_get_app_data_path(*args):
    shell32 = ctypes.WinDLL("shell32.dll")
    SHGetFolderPath = shell32.SHGetFolderPathW
    SHGetFolderPath.argtypes = (
        ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_uint32,
        ctypes.c_wchar_p)
    SHGetFolderPath.restype = ctypes.c_uint32

    CSIDL_LOCAL_APPDATA = 0x001c
    MAX_PATH = 260

    buf = ctypes.create_unicode_buffer(MAX_PATH)
    res = SHGetFolderPath(0, CSIDL_LOCAL_APPDATA, 0, 0, buf)
    if res != 0:
        raise Exception("Could not deterime APPDATA path")

    return os.path.join(buf.value, *args)


if __name__ == "__main__":
    if is_portable():
        win32_portable_main()
    else:
        main()
